### ISSUE TITLE
ASM: Allow multiple machines to access the same machine instance

### ISSUE THREAD

**Issue by georgwiese** [powdr-labs/powdr#1200]
We currently don't have a way to declare a machine instance that is used by more than one calling machine, like this:
```mermaid
graph TD;
    Main --> Hash;
    Main --> Padding;
    Padding --> Hash;
```

Automatic witness generation for this should already be working!

As far as I'm aware, this is the last missing features we'd need to implement #1055!

**Comment by Schaeff** [comment]
I think the cleanest way to achieve this would be to make instanciation explicit rather than implicit now:

```
let arith = Arith {};

let other = Other { arith }
let main = Main { other, arith }
```

And interpret the `main` instance as the entry point just like we use the `Main` machine type now.

**Comment by chriseth** [comment]
I agree, machines should be syntactic sugar for functions that return machine objects.

I think the main open question is where to place the arguments for machines (as in already instatiated sub-machines) if there are any.

**Comment by leonardoalt** [comment]
We decided in a call to do something like this:

```rust
machine Main {
    mem: Memory;
    binary: Binary(mem);
}

machine Binary(mem: Memory) {
    use mem.addr or whatever
    ...
    #[operation_id]
    col witness operation_id;
    
    #[latch]
    col fixed latch(i) { if (i % 4) == 3 { 1 } else { 0 } };

    ...
}

**Comment by leonardoalt** [comment]
Depends on https://github.com/powdr-labs/powdr/issues/1251

**Comment by leonardoalt** [comment]
Was this merged @pacheco ?

**Comment by georgwiese** [comment]
This has been solved for a while.
